assemble/Particle_Diagnostics.F90

```diff
@@ -0,0 +1,1558 @@
+
+    proc_num = getprocno()
+
+    !Check the ratio of particles present per group and determine which groups spawn
+    spawn_group = .true.
+    if (present(cap_percent)) then
+       spawn_group = .false.
+       do i=1,size(group_arrays)
+          if ((((node_particles(i)%length*1.0)/sum(node_particles(:)%length)*1.0)*100)>=cap_percent) then
+             spawn_group = .true.
+             group_spawn = i
+          end if
+       end do
+    end if
+
+    add_particles(:) = 0
+    !Return if no groups spawn
+    if (spawn_group.eqv..false.) return
+
+    !Count the number of particles per surrounding element for weighted spawning if
+    !not spawning based on radius from parent
+    if (have_radius.eqv..false.) then
+       ele_nums => node_neigh(xfield, node_num)
+       allocate(ele_num_part(size(ele_nums)))
+       ele_num_part(:) = 0
+       do j = 1,size(group_arrays)
+          particle => node_particles(j)%first
+          do while(associated(particle))
+             do i = 1,size(ele_nums)
+                if (ele_nums(i) == particle%element) then
+                   ele_num_part(i) = ele_num_part(i) + 1
+                end if
+             end do
+             particle => particle%temp_next
+          end do
+       end do
+    end if
+
+    !Loop over particle groups for spawning
+    do j = 1,size(group_arrays)
+       if (present(cap_percent)) then
+          if (j/=group_spawn) cycle
+       end if
+       !Calculated weighted attribute parameters based on existing particles if
+       !not copying parent attributes when spawning
+       if (copy_parents.eqv..false.) then
+          particle => node_particles(j)%first
+          if (associated(particle)) then
+             allocate(weighted_attributes(size(particle%attributes)))
+             allocate(weighted_old_attributes(size(particle%old_attributes)))
+             allocate(weighted_old_fields(size(particle%old_fields)))
+             weighted_attributes(:) = 0
+             weighted_old_attributes(:) = 0
+             weighted_old_fields(:) = 0
+          end if
+          do while(associated(particle))
+             weighted_attributes = weighted_attributes + particle%attributes
+             weighted_old_attributes = weighted_old_attributes + particle%old_attributes
+             weighted_old_fields = weighted_old_fields + particle%old_fields
+             particle => particle%temp_next
+          end do
+          if (allocated(weighted_attributes)) then
+             weighted_attributes = weighted_attributes/node_particles(j)%length
+             weighted_old_attributes = weighted_old_attributes/node_particles(j)%length
+             weighted_old_fields = weighted_old_fields/node_particles(j)%length
+          end if
+       end if
+
+       !Determine id, size and shape of spawned particle parameters
+       temp_part => particle_lists(group_arrays(j))%last
+       if (associated(temp_part)) then
+
+          id = particle_lists(group_arrays(j))%proc_part_count
```

Comment:
if temp_part isn't associated, we won't be using id anyway. temp_part is grabbing the last particle in the particle_list, so if it doesn't exist there won't be any particles in the list and
`particle => node_particles(j)%first`
will also be unassociated

Reply:
Okay, I think it was because `temp_part` comes from `particle_lists` vs. `node_particles` so I wasn't sure whether it could be unassociated but there still being something in `node_particles`. Your logic makes sense though, so it's fine!